File: mobile/src/crypto/mls.ts
/**
 * MLS-Layer — Platzhalter fuer Phase 3.
 *
 * In Phase 3 werden diese Funktionen durch Aufrufe in ein natives
 * Rust-Modul (OpenMLS via UniFFI) ersetzt, zusammen mit Arti als
 * eingebettetem Tor-Client. Die Signaturen hier entsprechen der
 * geplanten Schnittstelle, damit UI und API-Client bereits dagegen
 * gebaut werden koennen.
 */

export class NotImplementedError extends Error {
  constructor(feature: string) {
    super(`MLS.${feature} wird in Phase 3 implementiert (OpenMLS via UniFFI)`);
    this.name = 'NotImplementedError';
  }
}

export interface MLSKeyPackage {
  publicBlob: Uint8Array;
}

export interface MLSGroupHandle {
  groupId: string;
  epoch: number;
}

export async function generateKeyPackages(
  _identityPriv: Uint8Array,
  _count: number,
): Promise<MLSKeyPackage[]> {
  throw new NotImplementedError('generateKeyPackages');
}

export async function createGroup(
  _identityPriv: Uint8Array,
  _groupId: string,
): Promise<MLSGroupHandle> {
  throw new NotImplementedError('createGroup');
}

export async function inviteMembers(
  _group: MLSGroupHandle,
  _keyPackages: Uint8Array[],
): Promise<{ commit: Uint8Array; welcomes: Uint8Array[] }> {
  throw new NotImplementedError('inviteMembers');
}

export async function processWelcome(
  _identityPriv: Uint8Array,
  _welcomeBlob: Uint8Array,
): Promise<MLSGroupHandle> {
  throw new NotImplementedError('processWelcome');
}

export async function processCommit(
  _group: MLSGroupHandle,
  _commitBlob: Uint8Array,
): Promise<MLSGroupHandle> {
  throw new NotImplementedError('processCommit');
}

export async function encryptMessage(
  _group: MLSGroupHandle,
  _plaintext: Uint8Array,
): Promise<Uint8Array> {
  throw new NotImplementedError('encryptMessage');
}

export async function decryptMessage(
  _group: MLSGroupHandle,
  _ciphertext: Uint8Array,
): Promise<Uint8Array> {
  throw new NotImplementedError('decryptMessage');
}

File: native/secchat-core/src/mls.rs
use std::sync::Arc;

use crate::error::SecchatError;

#[derive(uniffi::Record, Clone)]
pub struct MlsKeyPackage {
    pub public_blob: Vec<u8>,
}

#[derive(uniffi::Record, Clone)]
pub struct MlsGroupHandle {
    pub group_id: String,
    pub epoch: u64,
}

#[derive(uniffi::Record, Clone)]
pub struct InviteOutcome {
    pub commit: Vec<u8>,
    pub welcomes: Vec<Vec<u8>>,
}

/// `MlsClient` haelt Identity- und Gruppen-State. Phase-3a-Stub —
/// Phase 3b wired `openmls_rust_crypto::OpenMlsRustCrypto` als Provider
/// und persistiert den KeyStore via Secure-Storage-Callback ans Expo-Modul.
#[derive(uniffi::Object)]
pub struct MlsClient {
    _identity_priv: Vec<u8>,
}

#[uniffi::export]
impl MlsClient {
    /// Instanziiert einen Client aus dem 32-Byte-Identity-Private-Key
    /// (abgeleitet aus dem Recovery-Seed, siehe `derive_identity_private_key`).
    #[uniffi::constructor]
    pub fn new(identity_priv: Vec<u8>) -> Result<Arc<Self>, SecchatError> {
        if identity_priv.len() != 32 {
            return Err(SecchatError::InvalidInput {
                reason: format!(
                    "identity_priv must be 32 bytes, got {}",
                    identity_priv.len()
                ),
            });
        }
        Ok(Arc::new(Self {
            _identity_priv: identity_priv,
        }))
    }

    pub fn generate_key_packages(
        &self,
        _count: u32,
    ) -> Result<Vec<MlsKeyPackage>, SecchatError> {
        Err(SecchatError::NotImplemented {
            what: "MlsClient.generate_key_packages".into(),
        })
    }

    pub fn create_group(
        &self,
        _group_id: String,
    ) -> Result<MlsGroupHandle, SecchatError> {
        Err(SecchatError::NotImplemented {
            what: "MlsClient.create_group".into(),
        })
    }

    pub fn invite_members(
        &self,
        _group: MlsGroupHandle,
        _key_packages: Vec<Vec<u8>>,
    ) -> Result<InviteOutcome, SecchatError> {
        Err(SecchatError::NotImplemented {
            what: "MlsClient.invite_members".into(),
        })
    }

    pub fn process_welcome(
        &self,
        _welcome_blob: Vec<u8>,
    ) -> Result<MlsGroupHandle, SecchatError> {
        Err(SecchatError::NotImplemented {
            what: "MlsClient.process_welcome".into(),
        })
    }

    pub fn process_commit(
        &self,
        _group: MlsGroupHandle,
        _commit_blob: Vec<u8>,
    ) -> Result<MlsGroupHandle, SecchatError> {
        Err(SecchatError::NotImplemented {
            what: "MlsClient.process_commit".into(),
        })
    }

    pub fn encrypt_message(
        &self,
        _group: MlsGroupHandle,
        _plaintext: Vec<u8>,
    ) -> Result<Vec<u8>, SecchatError> {
        Err(SecchatError::NotImplemented {
            what: "MlsClient.encrypt_message".into(),
        })
    }

    pub fn decrypt_message(
        &self,
        _group: MlsGroupHandle,
        _ciphertext: Vec<u8>,
    ) -> Result<Vec<u8>, SecchatError> {
        Err(SecchatError::NotImplemented {
            what: "MlsClient.decrypt_message".into(),
        })
    }
}
